### PR TITLE
Ma/sustain 1052

### PR DESCRIPTION
**Pull request by markan**
Support for A2. 

This
* adds an a2-mode config flag
* modifies delivery setup to accept --a2-mode and create an appropriate config file
* and modifies resources to include /workflow when appropriate if --a2-mode is enabled.

### CODE REVIEW

File: src/delivery/cli/setup.rs [hunk 1]
```diff
@@ -57,20 +59,25 @@ impl<'n> SetupClapOptions<'n> {
...
             .set_server(&self.server)
             .set_user(&self.user)
             .set_enterprise(&self.ent)
             .set_organization(&self.org)
             .set_pipeline(&self.pipeline)
-            .set_project(&self.project);
-
+            .set_project(&self.project)
+            .set_a2_mode(self.a2_mode);
+        // A2 mode requires SAML right now
```
Comment (afiune) [inline src/delivery/cli/setup.rs]: So we can't use/login with local users in A2?

Comment (btm) [reply src/delivery/cli/setup.rs]: @afiune no, A2 doesn't expose an API that we can authenticate against with credentials yet. See the discussion of state in the comments of SHACK-393 on JIRA.

Comment (afiune) [reply src/delivery/cli/setup.rs]: Understood! 👍

File: src/delivery/config/mod.rs [hunk 2]
```diff
@@ -160,7 +187,18 @@ config_accessor_for!(
...
     "config_json not set; set it in your cli.toml"
 );
-config_accessor_for!(fips_git_port, set_fips_git_port, "You did not set the fips_git_port. Set this value in your cli.toml or pass --fips-git-port.\nIt should be set to any port that is free and open on localhost (i.e. `fips_git_port = \"36534\"` in your cli.toml).");
+config_accessor_for!(
+    fips_git_port,
+    set_fips_git_port,
+    "You did not set the fips_git_port. Set this value in your cli.toml or pass --fips-git-port.\nIt should be set to any port that is free and open on localhost (i.e. `fips_git_port = \"36534\"` in your cli.toml)."
+);
+
+config_bool_accessor_for!(
```
[the remark this answers is not in the document]
Comment (markan) [reply src/delivery/config/mod.rs]: We use the pattern 'config.set_foo().set_bar()....' a lot, and my thinking was having the accessor makes it easier to maintain parallel construction. Is there a better way to do that?

**Comment by btm** [comment]
Tests are being worked on here: https://github.com/chef/delivery-cli/pull/55

This builds and executes for me locally on my Ubuntu workstation. I haven't run it against an A2 Workflow server yet though.

**Comment by afiune** [comment]
Perfect! Real failures that we need to address.
```
Failing Scenarios:
--
  | cucumber features/setup.feature:13 # Scenario: setup with no additional data supplied
  | cucumber features/setup.feature:30 # Scenario: setup with enterprise
  | cucumber features/setup.feature:43 # Scenario: setup with organization
  | cucumber features/setup.feature:56 # Scenario: setup with user
  | cucumber features/setup.feature:69 # Scenario: setup with pipeline
  | cucumber features/setup.feature:81 # Scenario: setup with server
  | cucumber features/setup.feature:94 # Scenario: setup with project
  | cucumber features/setup.feature:107 # Scenario: setup with all the args
  | cucumber features/setup.feature:124 # Scenario: setup when a config file already exists
```

**Comment by afiune** [comment]
So, everything is green now! Which is very cool and clean. 

tenor-92895920

Now it is time to test that the changes actually enable us to use the delivery-cli in A2. 

I tried already to create a project and clone it but I didn't have so much luck. 
```
➜  workflow /Users/afiune/github/delivery-cli/target/release/delivery setup --ent=automate --org=local --user=afiune@chef.io --server=a2-dev.test --a2-mode
Chef Delivery
Writing configuration to /Users/afiune/workflow/.delivery/cli.toml
New configuration
-----------------
server = "a2-dev.test"
api_protocol = "https"
user = "afiune@chef.io"
enterprise = "automate"
organization = "local"
git_port = "8989"
pipeline = "master"
saml = true
a2_mode = true
➜  workflow /Users/afiune/github/delivery-cli/target/release/delivery clone awesome
Chef Delivery
Cloning ssh://afiune@chef.io@automate@a2-dev.test:8989/automate/local/awesome to awesome
Unable to clone project 'awesome'
Analyzing error:
Requesting Token
Press Enter to open a browser window to retrieve a new token.
Launching browser..
Enter token: eyJhbGciOiJSUzI1NiIsImtpZCI6ImNlZmRlZGM5NDUxOWVlYzUwMThhNjFiYWE0YjYxMWE4MmVmYjI2NDcifQ.eyJpc3MiOiJodHRwczovL2EyLWRldi50ZXN0L2RleCIsInN1YiI6IkNnWmhabWwxYm1VU0JITmhiV3ciLCJhdWQiOiJhdXRvbWF0ZS1zZXNzaW9uIiwiZXhwIjoxNTQzNDk2NTk4LCJpYXQiOjE1NDM0MTAxOTgsImF0X2hhc2giOiI4WTdKVUpXRGs1ZU1rZjJXalE3dXZRIiwiZW1haWwiOiJhZml1bmVAY2hlZi5pbyIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJncm91cHMiOlsiRXZlcnlvbmUiLCJBbGwgRGV2ZWxvcGVycyIsIkVudGlyZSBDb21wYW55IiwiQUQgQ2hlZiBBV1MiLCJDaGVmIFVTIl0sIm5hbWUiOiJhZml1bmVAY2hlZi5pbyIsImZlZGVyYXRlZF9jbGFpbXMiOnsiY29ubmVjdG9yX2lkIjoic2FtbCIsInVzZXJfaWQiOiJhZml1bmUifX0.lB2Xo7xGLabGvz2KFAu6f7ApWVz04g_5zsWkm341Naco7m1jB-PJRYiMiQrYp2FpLYD7B__v4R856X68Y5luWyW2dwLABap9lGzeSw-8VGDpYXfsY-QKrsJCn_QSbk20CebzqgyCxG7ZLAycTMtHp2hPjZCx009Cu8UJ_Zsx1yo2MJB0xvyT3F6sW-sOtapNxMC7fwsvpkEFjDESyQcyPgoMXmxPbq_8bIYDhgBFQ2MlOGEruKawMfBmVcCUfhVeVOBYpcbqQ7RZ3Ens-SltGOSKg96c1zzqacI8VD7JX-71sPqFw3h016Q-EkgWj9Wd4fl30_G3m6hkfqCdKk3VLg
token: eyJhbGciOiJSUzI1NiIsImtpZCI6ImNlZmRlZGM5NDUxOWVlYzUwMThhNjFiYWE0YjYxMWE4MmVmYjI2NDcifQ.eyJpc3MiOiJodHRwczovL2EyLWRldi50ZXN0L2RleCIsInN1YiI6IkNnWmhabWwxYm1VU0JITmhiV3ciLCJhdWQiOiJhdXRvbWF0ZS1zZXNzaW9uIiwiZXhwIjoxNTQzNDk2NTk4LCJpYXQiOjE1NDM0MTAxOTgsImF0X2hhc2giOiI4WTdKVUpXRGs1ZU1rZjJXalE3dXZRIiwiZW1haWwiOiJhZml1bmVAY2hlZi5pbyIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJncm91cHMiOlsiRXZlcnlvbmUiLCJBbGwgRGV2ZWxvcGVycyIsIkVudGlyZSBDb21wYW55IiwiQUQgQ2hlZiBBV1MiLCJDaGVmIFVTIl0sIm5hbWUiOiJhZml1bmVAY2hlZi5pbyIsImZlZGVyYXRlZF9jbGFpbXMiOnsiY29ubmVjdG9yX2lkIjoic2FtbCIsInVzZXJfaWQiOiJhZml1bmUifX0.lB2Xo7xGLabGvz2KFAu6f7ApWVz04g_5zsWkm341Naco7m1jB-PJRYiMiQrYp2FpLYD7B__v4R856X68Y5luWyW2dwLABap9lGzeSw-8VGDpYXfsY-QKrsJCn_QSbk20CebzqgyCxG7ZLAycTMtHp2hPjZCx009Cu8UJ_Zsx1yo2MJB0xvyT3F6sW-sOtapNxMC7fwsvpkEFjDESyQcyPgoMXmxPbq_8bIYDhgBFQ2MlOGEruKawMfBmVcCUfhVeVOBYpcbqQ7RZ3Ens-SltGOSKg96c1zzqacI8VD7JX-71sPqFw3h016Q-EkgWj9Wd4fl30_G3m6hkfqCdKk3VLg
saved API token to: /Users/afiune/.delivery/api-tokens
token: eyJhbGciOiJSUzI1NiIsImtpZCI6ImNlZmRlZGM5NDUxOWVlYzUwMThhNjFiYWE0YjYxMWE4MmVmYjI2NDcifQ.eyJpc3MiOiJodHRwczovL2EyLWRldi50ZXN0L2RleCIsInN1YiI6IkNnWmhabWwxYm1VU0JITmhiV3ciLCJhdWQiOiJhdXRvbWF0ZS1zZXNzaW9uIiwiZXhwIjoxNTQzNDk2NTk4LCJpYXQiOjE1NDM0MTAxOTgsImF0X2hhc2giOiI4WTdKVUpXRGs1ZU1rZjJXalE3dXZRIiwiZW1haWwiOiJhZml1bmVAY2hlZi5pbyIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJncm91cHMiOlsiRXZlcnlvbmUiLCJBbGwgRGV2ZWxvcGVycyIsIkVudGlyZSBDb21wYW55IiwiQUQgQ2hlZiBBV1MiLCJDaGVmIFVTIl0sIm5hbWUiOiJhZml1bmVAY2hlZi5pbyIsImZlZGVyYXRlZF9jbGFpbXMiOnsiY29ubmVjdG9yX2lkIjoic2FtbCIsInVzZXJfaWQiOiJhZml1bmUifX0.lB2Xo7xGLabGvz2KFAu6f7ApWVz04g_5zsWkm341Naco7m1jB-PJRYiMiQrYp2FpLYD7B__v4R856X68Y5luWyW2dwLABap9lGzeSw-8VGDpYXfsY-QKrsJCn_QSbk20CebzqgyCxG7ZLAycTMtHp2hPjZCx009Cu8UJ_Zsx1yo2MJB0xvyT3F6sW-sOtapNxMC7fwsvpkEFjDESyQcyPgoMXmxPbq_8bIYDhgBFQ2MlOGEruKawMfBmVcCUfhVeVOBYpcbqQ7RZ3Ens-SltGOSKg96c1zzqacI8VD7JX-71sPqFw3h016Q-EkgWj9Wd4fl30_G3m6hkfqCdKk3VLg
Verifying Token: valid
Attempted to parse invalid JSON
JSON error: invalid type: null, expected a string at line 7 column 15
```

I would like to approve this PR but I don't think we are 100% there. Should I help moving this forward? @markan @btm @krasnow @rhass

**Comment by btm** [comment]
@afiune what channel are you pulling from? do you have the latest `automate-workflow-server` that includes https://github.com/chef/automate/pull/1533? Can you include `RUST_LOG=debug` output?

This is working for me:

```
btm@btm-thinkpad:/tmp/delivery-test (initialize-delivery-pipeline)$ delivery setup --ent=demo --org=testing --user=admin --server=a2w --a2-mode
Chef Delivery
Writing configuration to /tmp/delivery-test/.delivery/cli.toml
New configuration
-----------------
server = "a2w"
api_protocol = "https"
user = "admin"
enterprise = "demo"
organization = "testing"
git_port = "8989"
pipeline = "master"
saml = true
a2_mode = true

btm@btm-thinkpad:/tmp/delivery-test (master)$ delivery init
Chef Delivery
Creating Delivery project...
  Skipping: Delivery project named delivery-test already exists.
Setting up the 'delivery' git remote...
  Skipping: The delivery git remote is up-to-date. (ssh://admin@demo@a2w:8989/demo/testing/delivery-test).
Pushing initial git history...
  Skipping: Found commits on remote for pipeline master, not pushing local commits.
Creating pipeline on Delivery server...
  Skipping: Delivery pipeline named master already exists for project delivery-test.
Generating build cookbook...
  Build cookbook generated at .delivery/build_cookbook.
  Build cookbook committed to git and pushed to pipeline named master.
Creating and committing DELIVERY.md readme...
  Feature branch named 'initialize-delivery-pipeline' created.
  DELIVERY.md created.
  DELIVERY.md committed in branch 'initialize-delivery-pipeline'.
Submitting feature branch 'initialize-delivery-pipeline' for review...
  Review submitted to Delivery.

Your new Delivery project is ready!
```

**Comment by afiune** [comment]
This is not working for SAML user but it does work for local users:
```
➜  awesome git:(master) /Users/afiune/github/delivery-cli/target/release/delivery init
Chef Delivery
Creating Delivery project...
  Skipping: Delivery project named awesome already exists.
Setting up the 'delivery' git remote...
  Skipping: The delivery git remote is up-to-date. (ssh://admin@automate@a2-dev.test:8989/automate/local/awesome).
Pushing initial git history...
  No git history found for pipeline master, pushing local commits from branch master.
Creating pipeline on Delivery server...
  Created Delivery pipeline master for project awesome.
Generating build cookbook...
  Build cookbook generated at .delivery/build_cookbook.
  Build cookbook committed to git and pushed to pipeline named master.
Creating and committing DELIVERY.md readme...
  Feature branch named 'initialize-delivery-pipeline' created.
  DELIVERY.md created.
  DELIVERY.md committed in branch 'initialize-delivery-pipeline'.
Submitting feature branch 'initialize-delivery-pipeline' for review...
  Review submitted to Delivery.

Your new Delivery project is ready!
```

I'll approve this change!

# SAML Error
We can't push to the internal git server because we can't modify the user details:
```
➜  local git:(master) /Users/afiune/github/delivery-cli/target/release/delivery init
Chef Delivery
Creating Delivery project...
  Skipping: Delivery project named local already exists.
Setting up the 'delivery' git remote...
  Skipping: The delivery git remote is up-to-date. (ssh://afiune@chef.io@automate@a2-dev.test:8989/automate/local/local).
Pushing initial git history...
Git command failed!
STDOUT:
STDERR: afiune@chef.io@automate@a2-dev.test: Permission denied (publickey).
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```
Server Error:
```
[2018-11-28T15:00:43+00:00] automate-workflow-server.default(O): 15:00:43.142 [error] Error in process <0.2051.0> on node 'delivery@127.0.0.1' with exit value:
[2018-11-28T15:00:43+00:00] automate-workflow-server.default(O): {[{reason,function_clause},{mfa,{deliv_hand_users_named,handle,2}},{stacktrace,[{deliv_user_json,translate_type,[<<"a2">>],[{file,"src/deliv_user_json.erl"},{line,34}]},{deliv_user_json,update_spec,2,[{file,"src/deliv_user_json.erl"},{line,60}]},{deliv_user_json,validate_update,2,[{file,"src/deliv_user_json.erl"},{line,24}]},{deliv_hand_users_named,do_handle,4,[{file,"src/deliv_hand_users_named.erl"},{line,146}]},{cowboy_rest,call,3,[{file,"src/cowboy_rest.erl"},{line,972}]},{cowboy_rest,process_content_type,3,[{file,"src/cowboy_rest.erl"},{line,773}]},{cowboy_protocol,execute,4,[{file,"src/cowboy_protocol.erl"},{line,435}]}]},{req,[{socket,#Port<0.6268>},{transport,ranch_tcp},{connection,close},{pid,<0.2051.0>},{method,<<"PUT">>},{version,'HTTP/1.1'},{peer,{{172,17,0,2},51156}},{host,<<"a2-dev.test">>},{host_info,undefined},{port,10301},{path,<<"/api/v0/e/automate/users/afiune@chef.io">>},{path_info,undefined},{qs,<<>>},{qs_vals,undefined},{bindings,[{user_name,<<"afiune@chef.io">>},{ent_name,<<"automate">>}]},{headers,[{<<"host">>,<<"a2-dev.test:10301">>},{<<"x-real-ip">>,<<"172.17.0.2">>},{<<"x-scheme">>,<<"https">>},{<<"x-forwarded-for">>,<<"172.17.0.1, 172.17.0.2">>},{<<"x-forwarded-proto">>,<<"https">>},{<<"connection">>,<<"close">>},{<<"content-length">>,<<"500">>},{<<"accept">>,<<"application/json, text/plain, */*">>},{<<"origin">>,<<"https://a2-dev.test">>},{<<"chef-delivery-token">>,<<"eyJhbGciOiJSUzI1NiIsImtpZCI6IjllMGU4Njg3MGI1NjM3ZjgzM2E2N2Y2NzE0MTQ5Njg4YzgwMmY0OTQifQ.eyJpc3MiOiJodHRwczovL2EyLWRldi50ZXN0L2RleCIsInN1YiI6IkNnWmhabWwxYm1VU0JITmhiV3ciLCJhdWQiOiJhdXRvbWF0ZS1zZXNzaW9uIiwiZXhwIjoxNTQzNTAzMTkwLCJpYXQiOjE1NDM0MTY3OTAsImF0X2hhc2giOiJzMGZCMHZHZjNMZTV6ajdlOFNZYVJnIiwiZW1haWwiOiJhZml1bmVAY2hlZi5pbyIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJncm91cHMiOlsiRXZlcnlvbmUiLCJBbGwgRGV2ZWxvcGVycyIsIkVudGlyZSBDb21wYW55IiwiQUQgQ2hlZiBBV1MiLCJDaGVmIFVTIl0sIm5hbWUiOiJhZml1bmVAY2hlZi5pbyIsImZlZGVyYXRlZF9jbGFpbXMiOnsiY29ubmVjdG9yX2lkIjoic2FtbCIsInVzZXJfaWQiOiJhZml1bmUifX0.Y1q5bzzlhvRBPsvXLwk7jF4kLfQpdBWWfzLdrUxaHOxWmDQoqGB412hIGm1vxfCMFwa4-isDQkE1EX1pxx86qgb5GTOWr6S5BGuQVQKASGcteJDtsJ3_UohtIjvRwZeH6lgIeU5R8lnLWGPXkmVdk194wjivspjCIvnuLCcj55mVzTAnKGaD-aWSEKh47judf0TD623IBVEOEhV8N29aEZHfcBlSoIaaN9sacFKIxnU9pLzEq7JRaMRLnkfgVXdA81SwkElgziM5xYTRSov6kahZkFeoHD41bguP80d76OfOk_b-6PTHwu58skrU_bEiqZK3qUgazx_ejqwo9DIhQw">>},{<<"chef-delivery-user">>,<<"afiune@chef.io">>},{<<"user-agent">>,<<"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/70.0.3538.102 Safari/537.36">>},{<<"content-type">>,<<"application/json;charset=UTF-8">>},{<<"referer">>,<<"https://a2-dev.test/workflow/e/automate/">>},{<<"accept-encoding">>,<<"gzip, deflate, br">>},{<<"accept-language">>,<<"en-US,en;q=0.9,es;q=0.8,fr;q=0.7">>},{<<"cookie">>,<<"chef-delivery-user=afiune%40chef.io; chef-delivery-token=eyJhbGciOiJSUzI1NiIsImtpZCI6IjllMGU4Njg3MGI1NjM3ZjgzM2E2N2Y2NzE0MTQ5Njg4YzgwMmY0OTQifQ.eyJpc3MiOiJodHRwczovL2EyLWRldi50ZXN0L2RleCIsInN1YiI6IkNnWmhabWwxYm1VU0JITmhiV3ciLCJhdWQiOiJhdXRvbWF0ZS1zZXNzaW9uIiwiZXhwIjoxNTQzNTAzMTkwLCJpYXQiOjE1NDM0MTY3OTAsImF0X2hhc2giOiJzMGZCMHZHZjNMZTV6ajdlOFNZYVJnIiwiZW1haWwiOiJhZml1bmVAY2hlZi5pbyIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJncm91cHMiOlsiRXZlcnlvbmUiLCJBbGwgRGV2ZWxvcGVycyIsIkVudGlyZSBDb21wYW55IiwiQUQgQ2hlZiBBV1MiLCJDaGVmIFVTIl0sIm5hbWUiOiJhZml1bmVAY2hlZi5pbyIsImZlZGVyYXRlZF9jbGFpbXMiOnsiY29ubmVjdG9yX2lkIjoic2FtbCIsInVzZXJfaWQiOiJhZml1bmUifX0.Y1q5bzzlhvRBPsvXLwk7jF4kLfQpdBWWfzLdrUxaHOxWmDQoqGB412hIGm1vxfCMFwa4-isDQkE1EX1pxx86qgb5GTOWr6S5BGuQVQKASGcteJDtsJ3_UohtIjvRwZeH6lgIeU5R8lnLWGPXkmVdk194wjivspjCIvnuLCcj55mVzTAnKGaD-aWSEKh47judf0TD623IBVEOEhV8N29aEZHfcBlSoIaaN9sacFKIxnU9pLzEq7JRaMRLnkfgVXdA81SwkElgziM5xYTRSov6kahZkFeoHD41bguP80d76OfOk_b-6PTHwu58skrU_bEiqZK3qUgazx_ejqwo9DIhQw; chef-delivery-enterprise=automate; ajs_user_id=null; ajs_anonymous_id=%22488b8897-2669-427c-aadc-234de68e0d30%22; _ga=GA1.2.389392855.1537716563; _gid=GA1.2.493486398.1543410329; ajs_group_id=%2200000000000...">>}]},...]},...],...}
[2018-11-28T15:00:43+00:00] automate-workflow-server.default(O): 15:00:43.143 [error] Ranch listener http terminated in deliv_hand_users_named:handle/2 with reason: no function clause matching deliv_user_json:translate_type(<<"a2">>) line 34
```

**Comment by afiune** [comment]
screen shot 2018-11-28 at 4 03 22 pm
screen shot 2018-11-28 at 4 03 28 pm

**Pull request merged**